src/bin/dashboard_src/overview_screen.rs

```diff
@@ -199,6 +200,7 @@ impl OverviewScreen {
                                 own_overview_data.syncing=resp.syncing;
                                 own_overview_data.synced_balance = Some(resp.synced_balance);
                                 own_overview_data.is_mining = resp.is_mining;
+                                own_overview_data.confirmations = resp.confirmations;
```

Comment:
sure, I can make this change if you want, but I guess my thought is two-fold:

1. this could be considered premature optimization.  ie, why remove useful functionality before we actually need to? 
2. Perhaps there is a way to speed up this function so it is o(1).   more below.

Ok so my knowledge of levelDB is near zero so I'm flying a bit blind and so far I've just gone with what the code already does with regards to the `monitored_utxos` data.

The basic issue is we have a conceptual table `monitored_utxos` that is indexed and sorted by the field `confirmed_in_block`.  We have another field, `spent_in_block` without any index, but we wish to find the max value of this field.  To do so we must perform a costly loop through all items.

In a SQL database we would simply add an index to `spent_in_block` and then perform a query like `select max(spent_in_block) from monitored_utxos`.  Due to the index, this becomes very fast.

iiuc, the monitored_utxos are stored in LevelDB, and retrieved in rusty_wallet_database.rs with this:

```
        let monitored_utxos_storage = schema.new_vec::<u64, MonitoredUtxo>("monitored_utxos");
```

Ok, so my questions would be:
1. Is there a way we could add an index on `spent_in_block` in LevelDB and do something equivalent to the SQL query described above?
2. Alternatively, would it be possible to add a field to wallet state that stores the max `spent_in_block` height?   This would make the reads very fast, with a hopefully very small cost when initializing/updating state, and also very small amount of storage/mem.

Clearly I need to read up some on LevelDB, but for now I figure it is better to ask you these questions.

Reply:
TL;DR: It's my impression that `rusty-leveldb` is very fast as long as you don't use its iterator functionality. The only place we're doing that is in the endpoint for clearing all standings:

**Don't do this without having a very good reason**
`src/models/state/mod.rs`
```rust
pub async fn clear_all_standings_in_database(&self) {
        let mut peer_databases = self.net.peer_databases.lock().await;

        let mut dbiterator: RustyLevelDBIterator<IpAddr, PeerStanding> =
            peer_databases.peer_standings.new_iter(); // VERY SLOW!

        for (ip, _v) in dbiterator.by_ref() {
            let old_standing = peer_databases.peer_standings.get(ip);

            if old_standing.is_some() {
                peer_databases
                    .peer_standings
                    .put(ip, PeerStanding::default())
            }
        }
    }

```

**About indices**
The way I would do "indices" in level-db is by making a table that contains the relevant mapping. The `MonitoredUtxo` set is currently implemented as a list which you have to loop over to get relevant information, meaning that search times for many queries will be $O(N)$, where $N$ is the number of monitored UTXOs. My intuition and, to some extent, experience tells me that this is OK for up to probably a few thousand entries. To get faster lookup times, you could add a table that contains information about the monitored UTXOs.

We have this kind of indexing implemented for blocks through the the "block index" table where the key data type and value data type looks like this. 

```rust
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum BlockIndexKey {
    Block(Digest),       // points to block headers and file locations
    File(u32),           // points to file information
    Height(BlockHeight), // Maps from block height to list of blocks
    LastFile,            // points to last file used
    BlockTipDigest,      // points to block digest of most canonical block known
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum BlockIndexValue {
    Block(Box<BlockRecord>),
    File(FileRecord),
    Height(Vec<Digest>),
    LastFile(LastFileRecord),
    BlockTipDigest(Digest),
}
```

To get a feel for how they're used, you can have a look at the `write_block` method in `src/models/state/archival_state.rs`.

I'm thinking, though, that we can wait with implementing such a "index" table until we observe functions related to `MonitoredUtxo` being too slow?

I'm OK with including the "confirmations" number on the dashboard's overview screen like you suggest. But I would like to log how long that lookup takes, in case we do that. That lookup time could then be logged at the `Debug` level.

Reply:
Ok, cool.  Thx very much for the explanation!

> I'm thinking, though, that we can wait with implementing such a "index" table until we observe functions related to MonitoredUtxo being too slow?

Sure I think that's fine.   Mainly my concern has been about usage for parties that automate thousands of transactions.  For example exchanges, traders, or merchants.   Of course by the time that is happening, there should be more dev resources available, and eventually multiple wallet options as well.

So for this PR then I will plan to leave the loop that finds max(spent_in_block).  I will add a link in the code to these comments about an index and also log time spent in the fn that looks up balance confirmations as you suggest.

I'm still curious about the process of adding an index table, and so I *might* subsequently play around with that just for learning/understanding.  But there are plenty of other things to work on also.